cmd/ris-mirror/rismirror/router.go

```diff
@@ -52,17 +55,60 @@ func (r *Router) GetVRFs() []*vrf.VRF {
 }
 
 func (r *Router) addVRF(rd uint64, sources []*grpc.ClientConn) {
-	v := r.vrfRegistry.CreateVRFIfNotExists(vrf.RouteDistinguisherHumanReadable(rd), rd)
+	r.vrfsMu.Lock()
+	defer r.vrfsMu.Unlock()
+
+	// cleanup old ris clients of the vrf, before freshly initializing the vrf with new connections
+	r._stopVRF(rd)
 
-	r.vrfs[rd] = newVRF(v.IPv4UnicastRIB(), v.IPv6UnicastRIB())
+	v := r.vrfRegistry.CreateVRFIfNotExists(vrf.RouteDistinguisherHumanReadable(rd), rd)
+	r.vrfs[rd] = newVRF(v)
 
+	// init all ris clients freshly
 	for _, src := range sources {
-		r.connectVRF(rd, src, 4)
-		r.connectVRF(rd, src, 6)
+		rc4 := r._connectVRF(rd, src, 4)
+		rc6 := r._connectVRF(rd, src, 6)
+		// we cannot call v directly here, because the ris clients are only in the specific implementation of the vrf struct.
+		r.vrfs[rd].Clients = append(r.vrfs[rd].Clients, rc4, rc6)
+	}
+}
+
+func (r *Router) removeVRF(rd uint64) {
+	r.vrfsMu.Lock()
+	defer r.vrfsMu.Unlock()
+	// for removal, first stop all vrfs
+	r._stopVRF(rd)
+	// then we clean the map entries
+	vrf := r.vrfs[rd]
+	r.vrfRegistry.UnregisterVRF(vrf.vrf)
+	delete(r.vrfs, rd)
+}
+
+func (r *Router) dropAllVRFs() {
+	for _, vrf := range r.GetVRFs() {
+		r.removeVRF(vrf.RD())
+	}
+}
+
+func (r *Router) _stopVRF(rd uint64) {
+	// check if we even have an vrf for this rd
+	v := r.vrfs[rd]
+	if v == nil {
+		return
+	}
+	// first step: stop all ris clients
+	for _, rc := range v.Clients {
+		if rc != nil {
```

Comment:
I'm wondering how a nil client would end up in the slice?

Reply:
Oh yes, if we ask to stop a VRF we should maybe know if it already exists.
Let me quickly check where we write to this slice. It should not contain nil at all.

Reply:
On a second thought, `func (rism *RISMirror) RemoveTarget(rtrName string, vrfRD uint64) {` would make even more sense, possibly with adjusting this method to return an error (instead of just logging it). Thoughts about this?

Reply:
Yes. No error necessary here. The nil check can be removed after fixing the make (see my other comment below).